azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/pom/PomXmlUpdater.java

```diff
@@ -0,0 +1,243 @@
+            final String artifactId = XmlUtils.getChildValue("artifactId", element);
+            final String version = XmlUtils.getChildValue("version", element);
+            if (plugin.getGroupId().equals(groupId) && plugin.getArtifactId().equals(artifactId) &&
+                    (plugin.getVersion().equals(version) || version == null)) {
+                springPluginNode = element;
+            }
+        }
+        if (springPluginNode == null) {
+            springPluginNode = createMavenSpringPluginNode(pluginsNode);
+        }
+
+        final Element configurationNode = createToPath(springPluginNode, Arrays.asList("configuration"));
+        app.applyToDom4j(configurationNode);
+        final Element deployNode = createToPath(configurationNode, Arrays.asList("deployment"));
+        deploy.applyToDom4j(deployNode);
+        // newly created nodes are not LocationAwareElement
+        Element firstCreatedNode = configurationNode;
+        while (!(firstCreatedNode.getParent() instanceof LocationAwareElement)) {
+            firstCreatedNode = firstCreatedNode.getParent();
+        }
+        FileUtils.fileWrite(pom, formatElement(firstCreatedNode.getParent(), firstCreatedNode));
+    }
+
+    private static String formatElement(Element parent, Element self) {
+        final String[] ls = IndentUtil.splitLines(parent.getDocument().asXML());
+        final String baseIndent = IndentUtil.calcXmlIndent(ls, ((LocationAwareElement) parent).getLineNumber() - 1,
+                ((LocationAwareElement) parent).getColumnNumber() - 2);
+        final String placeHolder = String.format("@PLACEHOLDER_RANDOM_%s@", RandomUtils.nextLong());
+        final Text placeHolderNode = new DefaultText("\n" + placeHolder);
+        // replace target node to placeholder
+        parent.content().replaceAll(t -> t == self ? placeHolderNode : t);
+        self.setParent(null);
+        // remove all spaces before target node
+        XmlUtils.trimTextBeforeEnd(parent, placeHolderNode);
+        final String originalXml = parent.getDocument().asXML();
+
+        final String[] ts = IndentUtil.splitLines(XmlUtils.prettyPrintElementNoNamespace(self));
+        final String replacement = Arrays.stream(ts).map(t -> baseIndent + "    " + t).collect(Collectors.joining("\n")) + "\n" + baseIndent;
+        return originalXml.replace(placeHolder, replacement);
+    }
+
+    private Element createMavenSpringPluginNode(Element pluginsRootNode) throws IOException {
+        final Element result = new DOMElement("plugin");
+        XmlUtils.addDomWithKeyValue(result, "groupId", this.plugin.getGroupId());
+        XmlUtils.addDomWithKeyValue(result, "artifactId", this.plugin.getArtifactId());
+        XmlUtils.addDomWithKeyValue(result, "version", this.plugin.getVersion());
+        pluginsRootNode.add(result);
+        return result;
+    }
+
+    private static Element createToPath(Element node, Iterable<String> paths) {
```

Comment:
change the parameters to `(Element node, String... path)`? So that no need to invoke `Arrays.asList` for this method

Reply:
fixed